wgsl/index.bs

```diff
@@ -4557,11 +4555,271 @@ TODO: *Stub*: Expression evaluation
+* Look at which vertices are reachable from "Value_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. Also note that the information that we need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Nothing}.
+
+TODO: do we need an extra behavior for possibly infinite loops? I should check whether some form of reconvergence happens after something like `while(tid mod 2) {}`.
+
+If the set of behaviors for a statement is a singleton, we take that to mean that reconvergence happened, and we pick the vertex corresponding to control-flow at the start of the statement as the vertex corresponding to control-flow at the exit of the statement, overriding the other rules below.
 
-### Divergence and reconvergence TODO ### {#divergence-reconvergence}
+The rules for analyzing statements take as argument both the statement itself and the vertex corresponding to control-flow at the beginning of it (which we'll note "CF" below) and return the following:
 
-### Uniformity restrictions TODO ### {#uniformity-restrictions}
+* A set of behaviors for the statement
+* A vertex corresponding to control-flow at the exit of it
+* A set of new vertices and edges to add to the graph
+
+In the table below, `CF1, S -> CF2, B` means "run the analysis on S starting with control-flow CF1, apply the required changes to the graph, and name the resulting control-flow CF2, and the resulting set of behaviors B".
+Similarly, `CF, E -> CF, V, B` means "run the analysis on expression E, starting with control-flow CF, apply the required changes to the graph, and name the resulting control-flow vertex CF, the resulting value vertex V, and the resulting set of behaviors B" (see next section for the analysis of expressions).
+
+We have a similar set of rules for expressions in left-value positions, that we denote by `CF, E => CF, L, B`. Instead of computing the vertex which corresponds to the uniformity of the value, it computes the vertex which corresponds to the uniformity of the memory location we are adressing.
+
+<table class='data'>
+  <caption>Uniformity rules for statements</caption>
+  <thead>
+    <tr><th>Statement<th>New vertices<th>Precondition<th>Behaviors<th>Resulting control-flow vertex<th>New edges
+  </thead>
+  <tr><td class="nowrap">*s1*; *s2*
+      <td>
+      <td class = "nowrap">*CF*, *s1* -> *CF1*, *B1*<br>
+          *CF1*, *s1* -> *CF2*, *B2*
+      <td>*B1* \ {Nothing} U *B2*
+      <td>*CF2*
+      <td>
```

Comment:
How can there not be a new edge between *s1* and *s2* in *s1*; *s2*? I think I'm missing something fundamental here.

Reply:
Oops, there is a typo here: it should be `*CF1*, *s2* -> *CF2*, *B2*` on the second line, I accidentally repeated *s1* instead of *s2*.

I guess it did not help with understanding, sorry. To further clarify things:
- s1 and s2 are statements, not vertices in the graph, so it makes no sense to have edges between them.
- the first line says that CF1 is the vertex corresponding to the control-flow at the exit of s1.
- the second line says to then use it as the control-flow at the entrance of s2, and name the vertex corresponding to the control-flow afterwards CF2
- so we don't need to explicitly create a new edge here, because the rules for `*CF1*, *s2* -> *CF2*, *B2*` must already have made a path from *CF2* to *CF1*, since we cannot have uniform control-flow after *s2* if it was non-uniform before it.

Reply:
+1  The way I'm thinking of it is that each statement generates two nodes in the analysis graph, essentially a precondition and a post-condition.   So the two lines express   CF1===Post(s1) === Pre(s1)  and so the path in the analysis graph is CF -> CF1 -> CF2